### PR TITLE
feat: Use the blurhash in Files

### PR DESCRIPTION
**Pull request by artonge**
List view | Grid view
-- | --
Screenshot from 2024-08-29 14-30-30 | Screenshot from 2024-08-29 14-30-26

### CODE REVIEW

File: apps/files/src/components/FileEntry/FileEntryPreview.vue [hunk 1]
```diff
@@ -14,16 +14,19 @@
...
-			alt=""
-			class="files-list__row-icon-preview"
-			:class="{'files-list__row-icon-preview--loaded': backgroundFailed === false}"
-			loading="lazy"
-			:src="previewUrl"
-			@error="onBackgroundError"
-			@load="backgroundFailed = false">
+		<!-- Decorative images, should not be aria documented -->
+		<span v-else-if="previewUrl" class="files-list__row-icon-preview-container">
+			<canvas v-if="hasBlurhash && (backgroundFailed === true || !backgroundLoaded)" ref="canvas" class="files-list__row-icon-blurhash" />
```
Comment (susnux) [inline apps/files/src/components/FileEntry/FileEntryPreview.vue:19]: Canvas have weird semantics for screen readers (or the problem is there is no standard how to work with them).
So I would recommend:
1. Either add `role="img"` and an empty aria label
2. Or probably better: Add `aria-hidden="true"`

Comment (artonge) [reply apps/files/src/components/FileEntry/FileEntryPreview.vue:19]: Then please: https://github.com/nextcloud/photos/pull/2628 :D

File: apps/files/src/components/FileEntry/FileEntryPreview.vue [hunk 2]
```diff
@@ -107,6 +111,7 @@ export default Vue.extend({
 	data() {
 		return {
 			backgroundFailed: undefined as boolean | undefined,
+			backgroundLoaded: false as boolean,
```
Comment (susnux) [inline apps/files/src/components/FileEntry/FileEntryPreview.vue:114]:
```suggestion
			backgroundLoaded: false,
```

Not needed as `boolean` is inferred from `false`

File: apps/files/src/components/FileEntry/FileEntryPreview.vue [hunk 3]
```diff
@@ -206,24 +211,53 @@ export default Vue.extend({
...
+		drawBlurhash() {
+			const width = this.$refs.canvas.width
+			const height = this.$refs.canvas.height
+
+			const pixels = decode(this.source.attributes['metadata-blurhash'], width, height)
+
+			const ctx = this.$refs.canvas.getContext('2d')
+			const imageData = ctx.createImageData(width, height)
+			imageData.data.set(pixels)
+			ctx.putImageData(imageData, 0, 0)
```
Comment (susnux) [inline apps/files/src/components/FileEntry/FileEntryPreview.vue:260]: There are some type issues, so I would recommend to first cast the canvas ref properly and also check for null on `getContext` (this is a valid return value):

```suggestion
		drawBlurhash() {
			const canvas = this.$refs.canvas as HTMLCanvasElement
			const width = canvas.width
			const height = canvas.height

			const pixels = decode(this.source.attributes['metadata-blurhash'], width, height)

			const ctx = canvas.getContext('2d')
			if (ctx === null) {
				logger.error('Cannot create context for blurhash canvas')
				return
			}
			const imageData = ctx.createImageData(width, height)
			imageData.data.set(pixels)
			ctx.putImageData(imageData, 0, 0)
```

**Review by susnux — COMMENTED** [review]
Looks super nice! Just a few a11y and TS comments :)

File: apps/files/src/components/FileEntry/FileEntryPreview.vue [hunk 4]
```diff
@@ -14,16 +14,19 @@
...
-			alt=""
-			class="files-list__row-icon-preview"
-			:class="{'files-list__row-icon-preview--loaded': backgroundFailed === false}"
-			loading="lazy"
-			:src="previewUrl"
-			@error="onBackgroundError"
-			@load="backgroundFailed = false">
+		<!-- Decorative images, should not be aria documented -->
+		<span v-else-if="previewUrl" class="files-list__row-icon-preview-container">
+			<canvas v-if="hasBlurhash && (backgroundFailed === true || !backgroundLoaded)" ref="canvas" class="files-list__row-icon-blurhash" aria-hidden="true"/>
```
Comment (susnux) [inline apps/files/src/components/FileEntry/FileEntryPreview.vue:19]:
```suggestion
			<canvas v-if="hasBlurhash && (backgroundFailed === true || !backgroundLoaded)"
				ref="canvas"
				aria-hidden="true"
				class="files-list__row-icon-blurhash" />
```

Make ESLint happy

File: apps/files/src/components/FileEntry/FileEntryPreview.vue [hunk 5]
```diff
@@ -14,16 +14,19 @@
...
+			<img v-if="backgroundFailed !== true"
```
Comment (susnux) [inline apps/files/src/components/FileEntry/FileEntryPreview.vue:20]: Prevent that both are shown, or am I missing something here?

```suggestion
			<img v-else-if="backgroundFailed !== true"
```

Comment (artonge) [reply apps/files/src/components/FileEntry/FileEntryPreview.vue:20]: No, we need both at the beginning, else the preview is never loaded by the ``. The canvas will then be hidden when the preview is loaded.

**Review by susnux — APPROVED** [review]

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**